### DISCUSSION TITLE
Debug strategy on "ep0 out was already available" issue

### DISCUSSION THREAD

**Discussion by whowechina** [hathach/tinyusb#2610] [Q&A]
Hey, I design open source rhythm music arcade cons based on rp2040.
One of them implements HID (NKRO and Gamepad) and CDC interfaces. CDC serial port 0 is for stdio and CDC serial port 1 is for other purpose. Now I'm facing "ep0 out was already available" (printed on stdio before the system freezes) issue. (project: https://github.com/whowechina/chu_pico)
It happens mostly after second or third time the Windows (host) side opens the HID device.
I'm not sure if this issue causes system freeze, or a system freeze causes this issue.

I don't know where to start.
Any friend here give me some advices, thank you very much.

**Comment by pekkanikander** [comment]
I have now a test case that reliably causes this panic, on each try.  However, the code quite complex and in a not so consistent state, due to lots of (partially unrelated) refactoring etc. going on.  But it is open source, so I can create a branch with exactly the current state of code and write instructions on how to repeat the panic.

Is there any desire to debug this further?  Any benefit from such messy test case?

I'll try to create also a cleaner test case, but I'm not sure if I will succeed within the time that I can allocate on this.

**Comment by whowechina** [comment]
After I changed report scheme (from arbitrary 1000Hz to edge-triggering plus a slow arbitrary report), it (ep0 error) went away.在 2025年7月25日，18:46，Pekka-Ilmari Nikander ***@***.***> 写道：﻿
I have now a test case that reliably causes this panic, on each try.  However, the code quite complex and in a not so consistent state, due to lots of (partially unrelated) refactoring etc. going on.  But it is open source, so I can create a branch with exactly the current state of code and write instructions on how to repeat the panic.
Is there any desire to debug this further?  Any benefit from such messy test case?
I'll try to create also a cleaner test case, but I'm not sure if I will succeed within the time that I can allocate on this.
